src/util/fs-normalized.js

```diff
@@ -0,0 +1,170 @@
+  try {
+    await unlink(data.dest);
+    await copyFilePoly(data.src, data.dest, 0, data);
+  } finally {
+    if (cleanup) {
+      cleanup();
+    }
+  }
+};
+
+// Node 8.5.0 introduced `fs.copyFile` which is much faster, so use that when available.
+// Otherwise we fall back to reading and writing files as buffers.
+const copyFilePoly: (src: string, dest: string, flags: number, data: CopyFileAction) => Promise<void> = (
+  src,
+  dest,
+  flags,
+  data,
+) => {
+  if (fs.copyFile) {
+    return new Promise((resolve, reject) =>
+      fs.copyFile(src, dest, flags, err => {
+        if (err) {
+          reject(err);
+        } else {
+          fixTimes(undefined, dest, data).then(() => resolve()).catch(ex => reject(ex));
+        }
+      }),
+    );
+  } else {
+    return copyWithBuffer(src, dest, flags, data);
+  }
+};
+
+const copyWithBuffer: (src: string, dest: string, flags: number, data: CopyFileAction) => Promise<void> = async (
+  src,
+  dest,
+  flags,
+  data,
+) => {
+  // Use open -> write -> futimes -> close sequence to avoid opening the file twice:
+  // one with writeFile and one with utimes
+  const fd = await open(dest, 'w', data.mode);
+  try {
+    const buffer = await readFileBuffer(src);
+    await write(fd, buffer, 0, buffer.length);
+    await fixTimes(fd, dest, data);
+  } finally {
+    await close(fd);
+  }
+};
+
+// We want to preserve file timestamps when copying a file, since yarn uses them to decide if a file has
+// changed compared to the cache.
+// There are some OS specific cases here:
+// * On linux, fs.copyFile does not preserve timestamps, but does on OSX and Win.
+// * On windows, you must open a file with write permissions to call `fs.futimes`.
+// * On OSX you can open with read permissions and still call `fs.futimes`.
+async function fixTimes(fd: ?number, dest: string, data: CopyFileAction): Promise<void> {
+  const doOpen = fd === undefined;
+  let openfd: number = fd ? fd : -1;
+
+  if (disableTimestampCorrection === undefined) {
+    // if timestamps match already, no correction is needed.
+    // the need to correct timestamps varies based on OS and node versions.
+    const destStat = await lstat(dest);
+    disableTimestampCorrection = fileDatesEqual(destStat.mtime, data.mtime);
+  }
+
+  if (disableTimestampCorrection) {
+    return;
+  }
+
+  if (doOpen) {
+    try {
+      openfd = await open(dest, 'a', data.mode);
+    } catch (er) {
+      // file is likely read-only
+      try {
+        openfd = await open(dest, 'r', data.mode);
+      } catch (err) {
+        // We can't even open this file for reading.
+        return;
+      }
+    }
+  }
+
+  try {
+    if (openfd) {
+      await futimes(openfd, data.atime, data.mtime);
+    }
+  } catch (er) {
+    // If `futimes` throws an exception, we probably have a case of a read-only file on Windows.
+    // In this case we can just return. The incorrect timestamp will just cause that file to be recopied
```

Comment:
Can we verify that the file is read-only and throw or log this `er` if that is not the case?

Reply:
There is no logger passed down to the `fs` modules, so I didn't have any easy way of logging these issues. I was trying to avoid even bigger changes by passing the reporter down another level. I also didn't want `fixTimes()` to ever cause a yarn install to fail. If we can't fix the time stamps, it will effect performance a bit, but won't really break anything. It'll just work like it does currently where sometimes modules might get rebuilt every time (#932 ) but that's better than failing the install. If I can bubble up an error without failing the install, I'll try to add a line to the verbose log. Otherwise I'd rather not restructure the code any more than I already have.